Lock bootstrap (x.py) build directory

worldeva opened the pull request:
Closes #76661,  closes #80849, 
`x.py` creates a lock file at `project_root/lock.db`

r? @jyn514 , because he was one that told me about this~

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @jyn514 (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

jyn514 commented:
Happy to be responsible for this, but @Mark-Simulacrum will want to take a look too I'm sure.

jyn514 commented on src/bootstrap/flock.rs:
```diff
@@ -0,0 +1,206 @@
+//! Handles creating a file lock to avoid race conditions[1].
```
This is now the third flock implementation in-tree, there's also one for rustdoc ... maybe we should separate this out into a tiny crate or something? It would be hard to get cargo to use it since it's a submodule, but we could at least share it with rustdoc.

cc @rust-lang/cargo, do you have opinions on this? Would you be willing to separate the flock implementation into a different crate, and maybe put it on rust-lang/flock? I don't think there's any need to publish to crates.io but it would be nice to share at least between different tools.
cc @kinnison @rbtcollins, I'm sure rustup has its own implementation of flock too.

rbtcollins replied:
https://github.com/rust-lang/rustup/issues/988#issuecomment-582314389 and the top level summary are probably relevant.

tl;dr: don't want the cargo implementation, it is buggy; but also haven't gotten around to fixing the issue in rustup, which is complected with https://github.com/rust-lang/rustup/issues/2441 making this a little gnarly to think through.

jyn514 replied:
Ok. x.py itself probably doesn't need an airtight flock implementation, I wouldn't expect people to be using NFS for builds. I totally get that rustup does have to deal with those scenarios though.

Does rustup's flock implementation have any downsides compared to cargo's? Could we start from that as the shared codebase instead?

thomcc replied:
There's also fd-lock, which worked relatively well for me in the past (admittedly it was a small project). https://crates.io/crates/fd-lock/

That said, perhaps you need more control than it gives.

jyn514 reviewed (commented):
I found that this doesn't lock in bootstrap.py, but maybe that's ok - certainly I haven't seen any bugs reported about it.

```
$ x check
// terminal 1
downloading https://ci-artifacts.rust-lang.org/rustc-builds/db002a06ae9154a35d410550bc5132df883d7baa/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
############################################################################### 100.0%
extracting /home/joshua/src/rust/rust/build/cache/llvm-db002a06ae9154a35d410550bc5132df883d7baa-False/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz

// terminal 2
$ x build
Updating only changed submodules
Submodules updated in 0.02 seconds
downloading https://ci-artifacts.rust-lang.org/rustc-builds/db002a06ae9154a35d410550bc5132df883d7baa/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
###                                                                               4.9%^C
```

I don't feel qualified to review the flock code, but I also think I may not have to - I left a comment below.

I'm very happy with how this fixes the parallel builds problem :) thank you for tackling it!

Mark-Simulacrum commented:
So... this might be a little "out there", but it feels simpler than the solution this suggests and avoids duplicating code from Cargo. Both Python 2.7.x and Python 3 natively have sqlite built-in, as far as I can tell, via the sqlite3 module.

That means that we could likely implement this by "just" leveraging sqlite transactions to block. This is sort of overkill, but I'd expect it to work pretty well across any platform or filesystem, since sqlite itself is pretty battle tested.

It's also true that the python script currently manages a bit of state via pretty adhoc files (e.g., for the download stamps). I wouldn't change those today, there's not much point, but in the future it might make sense to leverage a more structured format that sqlite trivially gives us for this storage and without creating a bunch of files all over the place.

I'm curious to hear what folks think of this idea. I haven't actually tried a sample implementation, either, but I'd naively expect it to work out to at most a couple dozen lines of Python -- not too complex. Since sqlite comes pre-compiled for us through python, this seems a nice solution that avoids some of the hassle we'd otherwise run into. If and/or when we migrate away from the python wrapper, we can explore alternatives (such as flock proposed by this PR), but for now this would seem better. It also avoids any parallelism issues in the python script's downloads and such, which also seems good -- I'm not sold those are bug free today.

I do have mixed feelings about this so I could be convinced the impl in this PR is something we should just go with, but if the sqlite-based solution actually works and isn't too hard, then it seems preferable to me personally on the grounds of not having to maintain pretty low-level platform-specific code in bootstrap -- which is a place I'd really prefer to avoid that.

jyn514 commented:
Hi @worldeva - do you know when you'll get a chance to look at this again?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4270751191?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/94a3a8b97514cfcb0e870d80261fb86211b4bb92/checks/4270751191/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
14 |     path: PathBuf,
   |     ^^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
error: could not compile `bootstrap` due to previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml
Build completed unsuccessfully in 0:00:58
```

</details>

worldeva commented:
Sorry for the long delay for the tiny commit; life came up >,<
As per suggestion, I've removed all the stuff in bootstrap-proper and moved it over to a `sqlite3` based bit in `x.py`.

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -1254,6 +1270,7 @@ def bootstrap(help_triggered):
 def main():
     """Entry point for the bootstrap process"""
     start_time = time()
+    lock = acquire_lock()
```
I think this is slightly earlier than necessary; it would be nice to print the help message before spending time acquiring the lock. I think a good spot is just before `build.update_submodules()` in `bootstrap()`.

worldeva replied:
Moved!~

jyn514 commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,25 @@
+        print("Waiting for lock on database")
+        con = sqlite3.Connection("xlock.db", timeout=9999)
+        curs = con.cursor()
+        curs.execute("BEGIN EXCLUSIVE")
```
Does this need to do some sort of cleanup to release the lock? Or is exiting the process enough?

worldeva replied:
Dropping the cursor should automatically end the transaction.

Mark-Simulacrum replied:
Have we verified what happens when x.py is killed (kill -9, or similar)? I would expect that the lock is just indefinitely held?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4422999136?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/9cd99a87b805e5ffd325db959e6272c885968fae/checks/4422999136/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0786
Found 502 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/src/bootstrap/bootstrap.py:1244: trailing whitespace
some tidy checks failed



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:11
```

</details>

jyn514 reviewed (approved):
This looks good to me with the tidy lint fixed :)

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,25 @@
 
 from time import time
 
+def acquire_lock():
```
It'd be great to put a comment on this function, explaining why we're using sqlite to implement this, and the goal of the lock itself.

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,25 @@
+    except sqlite3.OperationalError:
+        del con
+        del curs
+        print("Waiting for lock on database")
```
I think database here is an internal implementation detail -- I would suggest a message closer to Cargo's:

>  Blocking waiting for file lock on the registry index

Maybe: "Blocking waiting for build directory lock".

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,25 @@
+        del con
+        del curs
+        print("Waiting for lock on database")
+        con = sqlite3.Connection("xlock.db", timeout=9999)
```
sqlite3 suggests the timeout is in milliseconds -- which means this would be just 10 seconds, quite low compared to typical rustc builds. Do we have some evidence for it being in seconds or similar? Otherwise, I would suggest passing a larger value here -- probably aiming at ~1 hour or so, though it's not clear to me why we *need* a timeout -- don't we want to block indefinitely? (Can we specify that?)

worldeva replied:
Unfortunately, in my testing, `float('inf')` doesn't work (just instantly times out), so it has to be some smaller value.
However, the sqlite3 lib for python does timeout in seconds ("The default for the timeout parameter is 5.0 (five seconds).")

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,25 @@
+        curs = con.cursor()
+        curs.execute("BEGIN EXCLUSIVE")
+        return curs
+    except sqlite3.OperationalError:
```
Can we document in a comment what results in this error? I presume it's a failure to acquire the lock within the (nonexistent) timeout?

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,25 @@
 
+def acquire_lock():
+    try:
+        con = sqlite3.Connection("xlock.db", timeout=0)
```
Is this creating the database in the working directory? I think we probably want it in build/ -- build.build_dir probably?

worldeva replied:
I changed it to creating it in `build.rust_dir`: The lock locks a lot more than just the stuff in `build` (submodules, etc), and the folder may not exist at the time of lock creation.

Mark-Simulacrum commented:
Mostly just some documentation, but I expect we can move ahead once that's added. It'd be good to do some spot checking locally to confirm this mostly works, and make a note in the documentation (and PR description) the location of the database file, so if there's some bug with this it's easy to delete it.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4439429647?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/79022f6cb7bf43a96979a221e193b7387ec81e95/checks/4439429647/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: 
configure: run `python /checkout/x.py --help`
configure: 
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap.py", line 25, in acquire_lock
    con = sqlite3.Connection(path, timeout=0)
sqlite3.OperationalError: unable to open database file

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
Traceback (most recent call last):
  File "../x.py", line 27, in <module>
    bootstrap.main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1294, in main
    bootstrap(help_triggered)
  File "/checkout/src/bootstrap/bootstrap.py", line 1256, in bootstrap
    lock = acquire_lock(build.rust_root)
  File "/checkout/src/bootstrap/bootstrap.py", line 35, in acquire_lock
    del con
UnboundLocalError: local variable 'con' referenced before assignment
```

</details>

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,34 @@
+# we don't cause a race condition while building
+# Lock is created in `project_root/lock.db`
+def acquire_lock(root_dir):
+    path = os.path.join(root_dir, "lock.db")
```
I'm not sure creating lock.db in the root is OK -- it feels a little messy. I think we should be able to create the build directory earlier, if it's not created yet, and place lock.db inside it.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4449852964?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/af6b3b226cdef7eca5f33a1e5051f197e136f985/checks/4449852964/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: 
configure: run `python /checkout/x.py --help`
configure: 
Traceback (most recent call last):
  File "../x.py", line 27, in <module>
    bootstrap.main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1295, in main
    bootstrap(help_triggered)
  File "/checkout/src/bootstrap/bootstrap.py", line 1256, in bootstrap
    os.mkdirs(build.build_dir, exist_ok=True)
AttributeError: module 'os' has no attribute 'mkdirs'
```

</details>

worldeva commented:
Alright!~
Testing shows that the lock is closed even when `SIGTERM`'d: so we don't have to worry about that :D
One thing I am worried about is that it won't exit from Ctrl-C while blocking (sqlite3 seems to catch `KeyboardInterrupt`?)

Mark-Simulacrum commented:
I think it's worth trying to investigate that more, seems like a bad thing -- maybe we can workaround that some how?

worldeva commented:
Alright, fixed: it seems like sqlite3 was blocking on the C side, meaning that python could never handle any signals.

Mark-Simulacrum commented:
@bors r+ 

Squashed the commits, let's land this; if there's problems we can always back this out.

bors commented:
:pushpin: Commit 0a84930b92b7bfe1155f22fe36e0f7c9dbdb6ff5 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 0a84930b92b7bfe1155f22fe36e0f7c9dbdb6ff5 -->
<!-- homu: {"type":"Approved","sha":"0a84930b92b7bfe1155f22fe36e0f7c9dbdb6ff5","approver":"Mark-Simulacrum"} -->

matthiaskrgr commented:
@bors r-

https://github.com/rust-lang/rust/pull/91664#issuecomment-988892478
Failed in a rollup.

matthiaskrgr commented on src/bootstrap/bootstrap.py:
```diff
@@ -12,9 +12,38 @@
 import sys
 import tarfile
 import tempfile
+import sqlite3
 
```
Maybe the sqlite3 python package is not available on all CI build runners...?

Mark-Simulacrum replied:
I think this particular feature would be fine to skip running if we don't have sqlite3 available. My hope would be that it's mostly available in practice -- so we should still have good support on practical usage, where we might expect parallel invocations.

worldeva commented:
Should be fixed!~

Mark-Simulacrum commented on src/bootstrap/bootstrap.py:
```diff
@@ -15,6 +15,40 @@
+                    pass
+            return curs
+    except ImportError:
+        print("warning: sqlite3 not available, skipping lock")
```
I think it's worth asking users to file an issue if they see this (on rust-lang/rust) and making it clear that the lock isn't necessary for correctness of the results (something like "build directory lock", maybe)?

worldeva commented:
Done-
The one thing that I'm mildly worried about is `"consider filing a issue"` makes it seem more important than it is- should I perhaps make it even more explicit with a "this lock is only to prevent multiple concurrent bootstraps from running" or such?
(...Though we're straying into the bikeshedding territory now :p )

Mark-Simulacrum commented:
Slightly adjusted wording, but seems good to me otherwise. Thanks!

@bors r+

bors commented:
:pushpin: Commit 1326bd6484725ed78552cd4b7ed7b5ddb5b5eec9 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 1326bd6484725ed78552cd4b7ed7b5ddb5b5eec9 -->
<!-- homu: {"type":"Approved","sha":"1326bd6484725ed78552cd4b7ed7b5ddb5b5eec9","approver":"Mark-Simulacrum"} -->
